DEFI/LiquidityPool/certora/specs/Full.spec

```diff
@@ -0,0 +1,448 @@
+*/
+rule reachability(method f)
+{
+	env e;
+	calldataarg args;
+	f(e,args);
+	satisfy true;
+}
+
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ The total shares supply of the system is less than equal the underlying asset holding of the system                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant totalSharesLessThanUnderlyingBalance()
+    totalSupply() <= underlying.balanceOf(currentContract)
+    {
+        preserved with(env e) {
+            require e.msg.sender != currentContract;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ The total shares supply of the system is zero if and only if the underlying asset holding of the system is zero    │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant totalSharesIsZeroWithUnderlying()
+		totalSupply() == 0 <=> underlying.balanceOf(currentContract) == 0
+		{
+			preserved with(env e) {
+				require e.msg.sender != currentContract;
+			}
+		}
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Sum of shares equal total shares supply                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+ghost mathint sumOfShares {
+    init_state axiom sumOfShares == 0;
+}
+
+hook Sstore _balanceOf[KEY address user] uint256 newSharesBalance (uint256 oldSharesBalance) STORAGE
+{
+    sumOfShares = sumOfShares + newSharesBalance - oldSharesBalance;
+}
+
+invariant totalSharesEqualSumOfShares()
+		to_mathint(totalSupply()) == sumOfShares;
+
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Sum of underlying equal total supply                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+ghost mathint sumBalances {
+    init_state axiom sumBalances == 0;
+}
+
+hook Sstore underlying._balanceOf[KEY address user] uint256 newBalance (uint256 oldBalance) STORAGE
+{
+    sumBalances = sumBalances + newBalance - oldBalance;
+}
+
+invariant totalIsSumBalances()
+    to_mathint(underlying.totalSupply()) == sumBalances;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Client shares and balance can (increase and decrease) or (decrease and increase)                                    │
```

Comment:
we call this anti-monotonicity (if one increases the other decreases )

Reply:
Done